<a href="https://colab.research.google.com/github/tatianaportsova/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Copy_of_LS_DS_123_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 123

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate):` 

You should only truly need these two values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?

### 2) Explore `scipy.stats.bayes_mvs`  
Read its documentation, and experiment with it on data you've tested in other ways earlier this week.
 - Create a visualization comparing the results of a Bayesian approach to a traditional/frequentist approach. (with a large sample size they should look close to identical, however, take this opportunity to practice visualizing condfidence intervals in general. The following are some potential ways that you could visualize confidence intervals on your graph:
  - [Matplotlib Error Bars](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.errorbar.html)
  - [Seaborn barplot with error bars](https://seaborn.pydata.org/generated/seaborn.barplot.html)
  - [Vertical ines to show bounds of confidence interval](https://www.simplypsychology.org/confidence-interval.jpg)
  - [Confidence Intervals on Box Plots](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.boxplot.html)

### 3) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).



### Using Bayes Theorem Iteratively (repeated testing)

![Bayes Theorem Drug Test Example](https://wikimedia.org/api/rest_v1/media/math/render/svg/95c6524a3736c43e4bae139713f3df2392e6eda9)

In [0]:
highest_p = (0,0,'')
for col in df.columns:
  if col == 'party':
    continue
  test = ttest_ind(dem1[col], rep1[col])
  if test[1] > highest_p[1]:
    highest_p = (test[0], test[1], col)
highest_p

In [0]:
def prob_drunk_given_positive(prob_drunk_prior = 1/1000, true_positive_rate = 1, false_positive_rate = .08): 
  p_nonuser = 1 - prob_drunk_prior
  
prior = None
for _ in range(5):
    if prior:
        prior = drunk_driving(0.08, 1, prior)
        print(prior)
    else:
        prior = drunk_driving(0.08, 1, 0.001)


In [0]:
drunk_driving = (0.08, 1, 0.001)
prior = None
for _ in range(5):
    if prior:
        prior = drunk_driving(0.08, 1, prior)
        print(prior)
    else:
        prior = drunk_driving(0.08, 1, 0.001)

drunk_driving

In [0]:
  posterior_probability = (true_positive_rate*prob_drunk_prior) / ((true_positive_rate*prob_drunk_prior) + (false_positive_rate*p_nonuser))
  while posterior_probability < 1:
    new_posterior= (true_positive_rate*posterior_probability) / ((true_positive_rate*posterior_probability) + (false_positive_rate*p_nonuser))
    posterior_probability = new_posterior
    print(posterior_probability)

In [0]:
prob_drunk_given_positive()

In [2]:
# True Positive Rate 100%
p_pos_user = 1
# Prior Probability (1/1000)
p_user = 1/1000
# False Positive Rate 1%
p_pos_nonuser = .08
# Complement probability of p_user (1 - p_user) or 999/1000
p_nonuser = 1 - p_user

print('Probability Positive Given User', p_pos_user)
print('Probability User', p_user)
print('Probability Positive Given Non-user', p_pos_nonuser)
print('Probability Non-user', p_nonuser)

Probability Positive Given User 1
Probability User 0.001
Probability Positive Given Non-user 0.08
Probability Non-user 0.999


In [3]:
posterior_probability = (p_pos_user*p_user) / ((p_pos_user*p_user) + (p_pos_nonuser*p_nonuser))

# The probabiltiy of a person who tests positive *actually* being a user:
print("Posterior Probability", posterior_probability)

Posterior Probability 0.012357884330202669


In [4]:
# True Positive Rate 100%
p_pos_user = 1
# Posterior Probability
p_user = posterior_probability
# False Positive Rate 1%
p_pos_nonuser = 0.08
# Complement probability of p_user (1 - p_user) or 999/1000
p_nonuser = 1 - p_user

print('Probability Positive Given User', p_pos_user)
print('Probability User', p_user)
print('Probability Positive Given Non-user', p_pos_nonuser)
print('Probability Non-user', p_nonuser)

Probability Positive Given User 1
Probability User 0.012357884330202669
Probability Positive Given Non-user 0.08
Probability Non-user 0.9876421156697973


In [5]:
posterior_probability = (p_pos_user*p_user) / ((p_pos_user*p_user) + (p_pos_nonuser*p_nonuser))

# The probabiltiy of a person who tests positive *actually* being a user:
print("Posterior Probability", posterior_probability)

Posterior Probability 0.13525210993291495


In [0]:
# Iterable function to find highest prior
def prob_drunk_iterator(prob_drunk_prior=1/1000, prob_positive=.08, prob_positive_drunk=1):
  # Calculate posterior
  posterior = (prob_positive_drunk * prob_drunk_prior) / prob_positive
  # Posterior becomes new prior, iterate through the equation until probability is 1
  while posterior < 1:
    new_posterior = (prob_positive_drunk * posterior) / prob_positive
    posterior = new_posterior
    print(posterior)

In [7]:
# Answer is the one before exceeding 1 (highest posterior we can get)
prob_drunk_iterator()

0.15625
1.953125


**1.2% is the probability** that a person is drunk after one positive breathalyzer test.

**13.5% is the probability** that a person is drunk after two positive breathalyzer tests.

**4 positive breathalyzer tests** are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit (96%).

### 2) Explore `scipy.stats.bayes_mvs`  
Read its documentation, and experiment with it on data you've tested in other ways earlier this week.
 - Create a visualization comparing the results of a Bayesian approach to a traditional/frequentist approach. (with a large sample size they should look close to identical, however, take this opportunity to practice visualizing condfidence intervals in general. The following are some potential ways that you could visualize confidence intervals on your graph:
  - [Matplotlib Error Bars](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.errorbar.html)
  - [Seaborn barplot with error bars](https://seaborn.pydata.org/generated/seaborn.barplot.html)
  - [Vertical ines to show bounds of confidence interval](https://www.simplypsychology.org/confidence-interval.jpg)
  - [Confidence Intervals on Box Plots](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.boxplot.html)

In [0]:
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns

In [136]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-01-23 03:39:52--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-01-23 03:39:53 (129 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [323]:
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])
print(df.shape)

(435, 17)


In [0]:
df = df.replace({'?':np.NaN, 'n':0, 'y':1})

In [338]:
rep = df[df.party == "republican"]
print(rep.shape)
rep.head(3)

(168, 17)


party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
7  republican                  0.0  ...        NaN           1.0

[3 rows x 17 columns]

In [339]:
rep = rep.drop(['handicapped-infants', 'water-project', 'budget',
                  'physician-fee-freeze', 'el-salvador-aid', 'religious-groups',
                  'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 'immigration',
                  'synfuels', 'education', 'right-to-sue',
                  'crime', 'south-africa'], 
                                                axis=1)

rep.head()

party  duty-free
0   republican        0.0
1   republican        0.0
7   republican        NaN
8   republican        0.0
10  republican        0.0

In [0]:
rep = rep.dropna()

In [341]:
rep.describe()

duty-free
count  156.000000
mean     0.089744
std      0.286735
min      0.000000
25%      0.000000
50%      0.000000
75%      0.000000
max      1.000000

In [0]:
data = rep["duty-free"]

In [343]:
# Bayesian approach
bayesian_confidence_interval, _, _ = stats.bayes_mvs(data, alpha=0.5)
print(bayesian_confidence_interval)

Mean(statistic=0.08974358974358974, minmax=(0.07422282293124735, 0.10526435655593214))


In [0]:
# Traditional/frequentist approach
def confidence_interval(data, confidence=0.95):
  data = np.array(data)
  mean = np.mean(data)
  n = len(data)

  s = np.std(data, ddof=1) 
  stderr = s / np.sqrt(n)
  margin_of_error = stderr * stats.t.ppf((1 + confidence) / 2.0, n - 1)
  return (mean, mean - margin_of_error, mean + margin_of_error)

In [345]:
votes = data(n=1, p=.5, size=1000)

confidence_interval(votes)

TypeError: ignored

### 3) In your own words, summarize the difference between Bayesian and Frequentist statistics



 **Frequentist statistics:**  The sampling is endless, and decision-making rules can be precise. The data is a reproducible random sample - there is a frequency. The basic parameters are fixed, that is, they remain constant during this repeated sampling process.

 **Bayesian statistics:** unknown quantities are considered probabilistically, and the state of the world can always be updated. Data is observed from the realized sample. The parameters are unknown and described probabilistically. This is data that is fixed.

#Resources#




- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!